File: 0001-0100/0073.set-matrix-zeroes.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        const int R = matrix.size();
        if (R == 0) return;
        const int C = matrix[0].size();
        if (C == 0) return;
        // 如果每行左边有0标注, 说明这行需要被删掉
        // 如果第一行的第c列为0, 说明第c列要被删掉
        bool del_first_row = false;
        for (int c = 0; c < C; ++c) {
            if (matrix[0][c] == 0) {
                del_first_row = true;
                break;
            }
        }
        for (int r = 1; r < R; ++r) {
            for (int c = 0; c < C; ++c) {
                if (matrix[r][c] == 0) {
                    matrix[r][0] = 0;
                    matrix[0][c] = 0;
                }
            }
        }
        for (int r = 1; r < R; ++r) {
            // 需要逆序, 因为第一列可能被置零
            for (int c = C - 1; c >= 0; --c) {
                if (matrix[0][c] == 0 || matrix[r][0] == 0)
                    matrix[r][c] = 0;
            }
        }
        if (del_first_row) {
            for (int c = 0; c < C; ++c) matrix[0][c] = 0;
        }
    }
};

File: 0001-0100/0073-set-matrix-zeroes.rs
struct Solution {}
impl Solution {
    pub fn set_zeroes(matrix: &mut Vec<Vec<i32>>) {
        if matrix.is_empty() {
            return
        }
        let first_row_is_zero = matrix[0].iter().any(|x| *x == 0);
        for r in 1..matrix.len() {
            for c in 0..matrix[r].len() {
                if matrix[r][c] == 0 {
                    matrix[r][0] = 0;
                    matrix[0][c] = 0;
                }
            }
        }
        for r in 1..matrix.len() {
            for c in 1..matrix[r].len() {
                if matrix[r][0] == 0 || matrix[0][c] == 0 {
                    matrix[r][c] = 0;
                }
            }
        }
        if matrix[0][0] == 0 {
            for r in 1..matrix.len() {
                matrix[r][0] = 0;
            }
        }
        if first_row_is_zero {
            // 全设为0, 包括matrix[0][0]
            for c in 0..matrix[0].len() {
                matrix[0][c] = 0;
            }
        }
            
    }
}

fn main() {
    let mut matrix = vec![vec![1,1,1], vec![1,0,1], vec![1,1,1]];
    Solution::set_zeroes(&mut matrix);
    assert_eq!(matrix, vec![vec![1,0,1], vec![0,0,0], vec![1,0,1]]);

    let mut matrix = vec![vec![1,0]];
    Solution::set_zeroes(&mut matrix);
    assert_eq!(matrix, vec![vec![0, 0]]);
}
